### PR TITLE
[WIP] Bump Django requirement to 1.11

### LINKED ISSUE #614: Add Django 1.11 support for 0.3 series
Hey,
there were some changes in django.forms.widgets in django 1.11. Looks like there is something broken now.

```
Traceback:

Environment:


Request Method: GET
Request URL: http://10.10.6.53:8080/_edit/

Django Version: 1.11
Python Version: 2.7.12
Installed Applications:
['django.contrib.admin',
 'django.contrib.auth',
 'django.contrib.contenttypes',
 'django.contrib.sessions',
 'django.contrib.messages',
 'django.contrib.staticfiles',
 'django.contrib.sites',
 'django.contrib.humanize',
 'django_nyt',
 'mptt',
 'sekizai',
 'sorl.thumbnail',
 'wiki',
 'wiki.plugins.attachments',
 'wiki.plugins.notifications',
 'wiki.plugins.images',
 'wiki.plugins.macros']
Installed Middleware:
['django.middleware.security.SecurityMiddleware',
 'django.contrib.sessions.middleware.SessionMiddleware',
 'django.middleware.common.CommonMiddleware',
 'django.middleware.csrf.CsrfViewMiddleware',
 'django.contrib.auth.middleware.AuthenticationMiddleware',
 'django.contrib.messages.middleware.MessageMiddleware',
 'django.middleware.clickjacking.XFrameOptionsMiddleware']


Template error:
In template /usr/local/lib/python2.7/dist-packages/wiki/templates/wiki/includes/formfield.html, error at line 14
   build_attrs() got an unexpected keyword argument 'name'   4 : 
   5 : <div id="div_{{ field.auto_id }}" class="form-group{% if field.errors %} has-error{% endif %}">
   6 :   <div class="wiki-label col-xs-3 col-lg-2">
   7 :   {% if field.label %}
   8 :     <label for="{{ field.id_for_label }}" class="control-label {% if field.field.required %}requiredField{% endif %}">
   9 :     {{ field.label|safe }}
   10 :     </label>
   11 :   {% endif %}
   12 :   </div>
   13 :   <div class="wiki-control col-xs-9 col-lg-10">
   14 :      {{ field }}  {% if field.field.required %}<span class="asteriskField">*</span>{% endif %}
   15 :     {% if field.errors %}
   16 :       {% for error in field.errors %}
   17 :         <div id="error_{{ forloop.counter }}_{{ field.auto_id }}" class="help-block"><strong>{{ error }}</strong></div>
   18 :       {% endfor %}
   19 :     {% endif %}
   20 :     {% if field.help_text %}
   21 :       <p id="hint_{{ field.auto_id }}" class="help-block">{{ field.help_text|safe }}</p>
   22 :     {% endif %}
   23 :   </div>
   24 : </div>


Traceback:

File "/usr/local/lib/python2.7/dist-packages/django/core/handlers/exception.py" in inner
  41.             response = get_response(request)

File "/usr/local/lib/python2.7/dist-packages/django/core/handlers/base.py" in _get_response
  217.                 response = self.process_exception_by_middleware(e, request)

File "/usr/local/lib/python2.7/dist-packages/django/core/handlers/base.py" in _get_response
  215.                 response = response.render()

File "/usr/local/lib/python2.7/dist-packages/django/template/response.py" in render
  107.             self.content = self.rendered_content

File "/usr/local/lib/python2.7/dist-packages/django/template/response.py" in rendered_content
  84.         content = template.render(context, self._request)

File "/usr/local/lib/python2.7/dist-packages/django/template/backends/django.py" in render
  66.             return self.template.render(context)

File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in render
  207.                     return self._render(context)

File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in _render
  199.         return self.nodelist.render(context)

File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in render
  990.                 bit = node.render_annotated(context)

File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in render_annotated
  957.             return self.render(context)

File "/usr/local/lib/python2.7/dist-packages/django/template/loader_tags.py" in render
  177.             return compiled_parent._render(context)

File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in _render
  199.         return self.nodelist.render(context)

File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in render
  990.                 bit = node.render_annotated(context)

File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in render_annotated
  957.             return self.render(context)

File "/usr/local/lib/python2.7/dist-packages/django/template/loader_tags.py" in render
  177.             return compiled_parent._render(context)

File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in _render
  199.         return self.nodelist.render(context)

File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in render
  990.                 bit = node.render_annotated(context)

File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in render_annotated
  957.             return self.render(context)

File "/usr/local/lib/python2.7/dist-packages/django/template/loader_tags.py" in render
  177.             return compiled_parent._render(context)

File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in _render
  199.         return self.nodelist.render(context)

File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in render
  990.                 bit = node.render_annotated(context)

File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in render_annotated
  957.             return self.render(context)

File "/usr/local/lib/python2.7/dist-packages/classytags/core.py" in render
  153.         return self.render_tag(context, **kwargs)

File "/usr/local/lib/python2.7/dist-packages/sekizai/templatetags/sekizai_tags.py" in render_tag
  93.         rendered_contents = nodelist.render(context)

File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in render
  990.                 bit = node.render_annotated(context)

File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in render_annotated
  957.             return self.render(context)

File "/usr/local/lib/python2.7/dist-packages/django/template/loader_tags.py" in render
  72.                 result = block.nodelist.render(context)

File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in render
  990.                 bit = node.render_annotated(context)

File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in render_annotated
  957.             return self.render(context)

File "/usr/local/lib/python2.7/dist-packages/django/template/loader_tags.py" in render
  72.                 result = block.nodelist.render(context)

File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in render
  990.                 bit = node.render_annotated(context)

File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in render_annotated
  957.             return self.render(context)

File "/usr/local/lib/python2.7/dist-packages/django/template/loader_tags.py" in render
  72.                 result = block.nodelist.render(context)

File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in render
  990.                 bit = node.render_annotated(context)

File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in render_annotated
  957.             return self.render(context)

File "/usr/local/lib/python2.7/dist-packages/django/template/defaulttags.py" in render
  529.             return self.nodelist.render(context)

File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in render
  990.                 bit = node.render_annotated(context)

File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in render_annotated
  957.             return self.render(context)

File "/usr/local/lib/python2.7/dist-packages/django/template/loader_tags.py" in render
  216.                 return template.render(context)

File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in render
  209.                 return self._render(context)

File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in _render
  199.         return self.nodelist.render(context)

File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in render
  990.                 bit = node.render_annotated(context)

File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in render_annotated
  957.             return self.render(context)

File "/usr/local/lib/python2.7/dist-packages/django/template/library.py" in render
  245.         return t.render(new_context)

File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in render
  209.                 return self._render(context)

File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in _render
  199.         return self.nodelist.render(context)

File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in render
  990.                 bit = node.render_annotated(context)

File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in render_annotated
  957.             return self.render(context)

File "/usr/local/lib/python2.7/dist-packages/django/template/defaulttags.py" in render
  216.                     nodelist.append(node.render_annotated(context))

File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in render_annotated
  957.             return self.render(context)

File "/usr/local/lib/python2.7/dist-packages/django/template/loader_tags.py" in render
  216.                 return template.render(context)

File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in render
  209.                 return self._render(context)

File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in _render
  199.         return self.nodelist.render(context)

File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in render
  990.                 bit = node.render_annotated(context)

File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in render_annotated
  957.             return self.render(context)

File "/usr/local/lib/python2.7/dist-packages/django/template/defaulttags.py" in render
  322.                 return nodelist.render(context)

File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in render
  990.                 bit = node.render_annotated(context)

File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in render_annotated
  957.             return self.render(context)

File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in render
  1046.         return render_value_in_context(output, context)

File "/usr/local/lib/python2.7/dist-packages/django/template/base.py" in render_value_in_context
  1024.     value = force_text(value)

File "/usr/local/lib/python2.7/dist-packages/django/utils/encoding.py" in force_text
  78.                 s = six.text_type(s)

File "/usr/local/lib/python2.7/dist-packages/django/utils/html.py" in <lambda>
  376.         klass.__unicode__ = lambda self: mark_safe(klass_unicode(self))

File "/usr/local/lib/python2.7/dist-packages/django/forms/boundfield.py" in __str__
  41.         return self.as_widget()

File "/usr/local/lib/python2.7/dist-packages/django/forms/boundfield.py" in as_widget
  120.             **kwargs

File "/usr/local/lib/python2.7/dist-packages/wiki/editors/markitup.py" in render
  61.         final_attrs = self.build_attrs(attrs, name=name)

Exception Type: TypeError at /_edit/
Exception Value: build_attrs() got an unexpected keyword argument 'name'
```

### PR DESCRIPTION
**Pull request by benjaoming**
Adding Django 1.11 support in prep for 0.3

Fixes #614

### CODE REVIEW

**Comment by benjaoming** [comment]
This is supposed to fail, so will need to figure out why it doesn't. It seems that rendering `{{ field }}` in the template produces an exception that doesn't get caught in tests.

**Comment by benjaoming** [comment]
Thanks to @atombrella for pointing to the Django issue concerning this:

https://code.djangoproject.com/ticket/28095

**Comment by spookylukey** [comment]
I hit this issue and started work on fixing it.

IMO the issue of why the tests don't catch the problem is the most important for future maintenance of the project. The current tests use the Django test Client, which is really not very helpful for testing views in many cases, especially any views which involve POST interactions. Client based tests are both verbose (requiring you to specify every field that you are POSTing for example, even if some of them would have come from hidden fields), and poorly test the actual requirements of the page - which really are defined in terms of what happens when you view a page in the web browser, and take an action etc. - which is at a different level of abstraction.

For that reason, I started looking at writing good tests first. However, that itself threw up several issues. Before I continue, please could I have your feedback @benjaoming on the following approach?

I plan to use [django-functest](http://django-functest.readthedocs.io/en/latest/) to add better tests, rewriting some of the existing ones as far as time will allow. This library allows functional style tests for web pages with a high level API. They can then be executed in two ways:

1) At the HTTP level, with something that works like a non-javascript browser
2) Automating a real browser e.g. Firefox or Chrome, using the Selenium framework.

That advantage of the second is that it actually reflects what a user will typically see, including Javascript etc. The disadvantage is that the tests are extremely slow, and sometimes flakey. So, I'd propose doing both, and having a mode for running just the former, for quick test runs.

To do this, however, some bits of functionality would have to be changed, because things like the wiki editing view depends on Javascript to work. IMO it does so unnecessarily, but to fix it would require changing the template HTML so that the form correctly wraps the form fields.

To summarise, I'm proposing a rewrite of the some of the view tests, with some significant changes to the views themselves to make them more testable. This is a risk, I understand, but in the end result you'll have tests that are significantly shorter, clearer, and much more realistic in what they test, including the ability to test in real web browsers as well as have fast tests that are not so accurate but still pretty good (and all of this from only one set of tests).

**Comment by benjaoming** [comment]
Hi @spookylukey 

Thanks for jumping in on this and for thoroughly explaining it!

I agree that there are issues about how the Django's test client does things, certainly this uncaught exception in template rendering shows that there are issues.

I also agree that writing more explicitly functional view tests would be a good approach, and django-functest looks like it has a nice and clean interface.

I have a sceptical attitude about Selenium because I've seen its downsides and struggled a lot with maintaining a bad codebase with loads of bad Selenium/BDD tests (that others wrote and I had to struggle with). Bad scenarios such as:

 1. Using (misusing) Selenium to test things that are inherently written in a badly testable way (as you correctly identify, the Edit page should be refactored so it's testable)
 1. Having tests and a test stack that's more complex than the code it's actually testing. I've seen too many cases of using functional browser tests with Selenium Just Because You Can (tm). So introducing it comes with the risk of accepting tests that are too complex (making them slow, buggy)
 1. Unstable tests and Selenium crashes - which are a real hit on productivity every time, you have to rebuild a CI instance because of false negatives.
 1. Tests that assert that things don't change rather than asserting that they work, making things harder to improve or maintain. Selenium tests often address a level of detail in the DOM that's subject to change. And so their test assumptions often break, and since these tests are so slow, fixing them also takes a substantial effort.
 1. Taking a negative hit on productivity because of test slowness

I'm certainly not implying that you would write bad tests :) I'm just afraid of the "creep" effect of introducing Selenium.

Do you think that you can achieve the kind of level of functional testing that you're advocating without Selenium? For instance if having django-functest and solely using the WebTest wrapper?

**Comment by benjaoming** [comment]
The other things you suggest sound good - in case you're pressed on time, perhaps it's possible to introduce django-functools in some smaller steps? Because having Django 1.11 is actually kind of urgent -- but I'm also pressed on time and not actually working on this PR right now.

**Comment by spookylukey** [comment]
@benjaoming totally on board with your reservations about Selenium etc.!

The way that django-functest works addresses some of your concerns - partly because it works at a high level, reducing the temptation to test the DOM directly etc.

With your reservations, I'd propose adding django-functest tests, including the ability to run Selenium versions, but in a normal test run (including CI), make only the WebTest wrapper run. This way, if you need to test the Javascript is still working, you can do so easily, but it won't slow you down or add any fragility for normal development lifecycle, and people will be encouraged to write code that is testable without Selenium.

In most cases, I should be able to achieve the kind of functional testing I'm looking for without Selenium. However, I'd need to change some templates (as mentioned before), and so it would be helpful for me to run the Selenium tests as well, to ensure that I haven't broken anything, as well as testing manually in a browser of course.

My initial tests suggest I can add the django-functest FuncWebTestMixin into an appropriate test base class without having to rewrite all the tests that use Django Client, which may be the best compromise at the moment.

**Comment by benjaoming** [comment]
Closing this in favour of @spookylukey 's PR #634- oh how code just communicates it all so well :)
